### PR TITLE
feat(nsqconsumer): log the error by using root context (ErrCtx)

### LINKED ISSUE #447: [NsqConsumer] Use root context to log error
Related to #442

### PR DESCRIPTION
**Pull request by curzolapierre**
- [x] Add a changelog entry in `CHANGELOG.md`
- [x] Rebase and update Errors dependency once https://github.com/Scalingo/go-utils/pull/449 is released

This PR in `nsqconsumer` package the implementation of ErrCtx to get context logger and be able to manage `noRetry` field correctly if the `Error` is wrapped

Fix #447

### CODE REVIEW

**Review by Soulou — CHANGES_REQUESTED** [review]
Once specs sorry but.. lack of specs :-D

File: nsqconsumer/CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # Changelog
 
+* feat: Start unwraps errors to find noRetry field that can be wrapped in ErrCtx. Also use ErrCtx to enrich the logger
+
 ## To be Released
```
Comment (EtienneM) [inline nsqconsumer/CHANGELOG.md:6]: ;)

```suggestion

## To be Released

* feat: Start unwraps errors to find noRetry field that can be wrapped in ErrCtx. Also use ErrCtx to enrich the logger
```

**Comment by curzolapierre** [comment]
To be able to add spec and test `Start` method, I need to mock `nsq.Consumer`, but we cannot mock it as one of called method (`SetLogger`) use an non exported param. Hence the new call to
```go
nsq.NewConsumer(consumer.Topic, consumer.Channel, consumer.NsqConfig)
```
won't work.

File: nsqconsumer/consumer.go [hunk 2]
```diff
@@ -167,84 +168,101 @@ func (c *nsqConsumer) Start(ctx context.Context) func() {
...
+		if errLogger == nil {
+			errLogger = msgLogger
 		}
-		return nil
-	}), c.MaxInFlight)
 
-	if err = consumer.ConnectToNSQLookupds(c.NsqLookupdURLs); err != nil {
-		c.logger.WithError(err).Fatalf("Fail to connect to NSQ lookupd")
+		if noRetry {
+			errLogger.WithError(err).Error("message handling error - noretry")
```
Comment (EtienneM) [inline nsqconsumer/consumer.go:255]: picky picky: while you are on it :)
```suggestion
			errLogger.WithError(err).Error("Message handling error - noretry")
```

File: nsqconsumer/consumer.go [hunk 3]
```diff
@@ -167,84 +168,101 @@ func (c *nsqConsumer) Start(ctx context.Context) func() {
...
+			return nil
+		}
+		errLogger.WithError(err).Error("message handling error")
```
Comment (EtienneM) [inline nsqconsumer/consumer.go:258]: picky picky: while you are on it :)
```suggestion
		errLogger.WithError(err).Error("Message handling error")
```

File: nsqconsumer/consumer.go [hunk 4]
```diff
@@ -167,84 +168,101 @@ func (c *nsqConsumer) Start(ctx context.Context) func() {
...
+
+	if len(message.Body) == 0 {
+		err := errgo.New("body is blank, re-enqueued message")
+		c.logger.WithError(err).Error("blank message")
+		return err
+	}
+	var msg NsqMessageDeserialize
+	err = json.Unmarshal(message.Body, &msg)
+	if err != nil {
+		c.logger.WithError(err).Error("failed to unmarshal message")
```
Comment (EtienneM) [inline nsqconsumer/consumer.go:207]: picky picky: while you are on it :)
```suggestion
		c.logger.WithError(err).Error("Fail to unmarshal message")
```

File: nsqconsumer/consumer.go [hunk 5]
```diff
@@ -167,84 +168,101 @@ func (c *nsqConsumer) Start(ctx context.Context) func() {
...
-			), "logger", msgLogger,
-		)
-
-		if msg.At != 0 {
-			now := time.Now().Unix()
-			delay := msg.At - now
-			if delay > 0 {
-				return c.postponeMessage(ctx, msgLogger, msg, delay)
+	if err = consumer.ConnectToNSQLookupds(c.NsqLookupdURLs); err != nil {
+		c.logger.WithError(err).Fatalf("Fail to connect to NSQ lookupd")
```
Comment (EtienneM) [inline nsqconsumer/consumer.go:174]: While we are on it, we shouldn't use `Fatalf`.

File: nsqconsumer/consumer.go [hunk 6]
```diff
@@ -167,84 +168,102 @@ func (c *nsqConsumer) Start(ctx context.Context) func() {
...
-				context.Background(), "request_id", msg.RequestID,
-			), "logger", msgLogger,
-		)
-
-		if msg.At != 0 {
-			now := time.Now().Unix()
-			delay := msg.At - now
-			if delay > 0 {
-				return c.postponeMessage(ctx, msgLogger, msg, delay)
+	if err = consumer.ConnectToNSQLookupds(c.NsqLookupdURLs); err != nil {
```
Comment (EtienneM) [inline nsqconsumer/consumer.go:173]: best practice: we don't use this construct, it's confusing

```suggestion
err = consumer.ConnectToNSQLookupds(c.NsqLookupdURLs)
	if err != nil {
```

**Review by EtienneM — CHANGES_REQUESTED** [review]

File: nsqconsumer/consumer.go [hunk 7]
```diff
@@ -167,84 +168,102 @@ func (c *nsqConsumer) Start(ctx context.Context) func() {
...
+	msgLogger := c.logger.WithFields(logrus.Fields{
+		"message_id":   fmt.Sprintf("%s", message.ID),
+		"message_type": msg.Type,
+		"request_id":   msg.RequestID,
+	})
 
-		before := time.Now()
-		if _, ok := c.SkipLogSet[msg.Type]; !ok {
-			msgLogger.Info("BEGIN Message")
+	//nolint:staticcheck,revive
```
Comment (EtienneM) [inline nsqconsumer/consumer.go:219]: You must explain why you add such a line

**Review by EtienneM — APPROVED** [review]

**Review by Soulou — APPROVED** [review]

**Pull request merged**